### PR TITLE
Explore: Fix auto completion on label values for Loki

### LINKED ISSUE #18695: Loki: label value suggestions are empty
0. Explore with Loki datasource
1. Expand the Label selector button to check for the list of available label names
2. Without selecting a label, enter the query field and start typing `{label=}` with your relevant label

Result: No suggestion after the `=`
Expected: Label values should be suggested.

Additional info:
After step 2 go back to the label selector button and expand your label. The suggestions are being loaded. Now go back to the `=` and tada, label values are being shown.

### PR DESCRIPTION
**Pull request by aocenas**
Closes: https://github.com/grafana/grafana/issues/18695

Fix was easy just added some initialisation but this contains lots of code changes:
- added tests
- during that I added some utility mocks and refactored other tests to use them
- added some typings and comments

### CODE REVIEW

File: public/app/plugins/datasource/loki/language_provider.test.ts [hunk 1]
```diff
@@ -83,23 +73,35 @@ describe('Language completion provider', () => {
...
-        },
-        { absoluteRange: rangeMock }
-      );
+      const input = createTypeaheadInput('{}', '');
+      const result = instance.provideCompletionItems(input, { absoluteRange: rangeMock });
       expect(result.context).toBe('context-labels');
       expect(result.suggestions).toEqual([{ items: [{ label: 'job' }, { label: 'namespace' }], label: 'Labels' }]);
     });
+
+    it('returns label suggestions from Loki', async () => {
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/language_provider.test.ts]: Added test for labels, which worked before but wasn't tested.

File: public/app/plugins/datasource/loki/language_provider.test.ts [hunk 2]
```diff
@@ -83,23 +73,35 @@ describe('Language completion provider', () => {
...
+      const datasource = makeMockLokiDatasource({ label1: [], label2: [] });
+      const provider = await getLanguageProvider(datasource);
+      const input = createTypeaheadInput('{}', '');
+      const result = provider.provideCompletionItems(input, { absoluteRange: rangeMock });
+      expect(result.context).toBe('context-labels');
+      expect(result.suggestions).toEqual([{ items: [{ label: 'label1' }, { label: 'label2' }], label: 'Labels' }]);
+    });
+
+    it('returns label values suggestions from Loki', async () => {
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/language_provider.test.ts]: This tests the actual fix.

File: public/app/plugins/datasource/loki/language_provider.test.ts [hunk 3]
```diff
@@ -226,3 +200,39 @@ describe('Labels refresh', () => {
...
+  await instance.start();
+  return instance;
+}
+
+/**
+ * @param value Value of the full input
+ * @param text Last piece of text (not sure but in case of {label=} this would be just '=')
+ * @param labelKey Label by which to search for values. Cutting corners a bit here as this should be inferred from value
+ */
+function createTypeaheadInput(
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/language_provider.test.ts]: This was a bit of guessing game for me as I wasn't sure what each of the params do, probably could be improved a bit so it does not need so much params.

File: public/app/plugins/datasource/loki/language_provider.ts [hunk 4]
```diff
@@ -54,7 +60,7 @@ export default class LokiLanguageProvider extends LanguageProvider {
   started: boolean;
   initialRange: AbsoluteTimeRange;
 
-  constructor(datasource: any, initialValues?: any) {
+  constructor(datasource: DataSourceApi, initialValues?: any) {
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/language_provider.ts]: I assume there is nothing else that can fit here except for `DataSourceApi` on the other hand we need `metadataRequest` there which actually isn't defined on `DataSourceApi` interface. So not ideal situation and I left `LanguageProvider` as it is with `this.datasource: any` (that is why `this.datasource.metadataRequest` call does not complain). Seems like I already added a lots of things to this PR, but we should probably make this more typesafe later.

File: public/app/plugins/datasource/loki/language_provider.ts [hunk 5]
```diff
@@ -74,21 +80,33 @@ export default class LokiLanguageProvider extends LanguageProvider {
...
     return this.startTask;
   };
 
-  // Keep this DOM-free for testing
-  provideCompletionItems({ prefix, wrapperClasses, text, value }: TypeaheadInput, context?: any): TypeaheadOutput {
+  /**
+   * Return suggestions based on input that can be then plugged into a typeahead dropdown.
+   * Keep this DOM-free for testing
+   * @param input
+   * @param context Is optional in types but is required in case we are doing getLabelCompletionItems
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/language_provider.ts]: The API here is a bit weird as `context` is optional and shaped based on what is in `wrapperClasses` which again isn't very typesafe. Another candidate for refactor I think.

File: public/app/plugins/datasource/loki/language_provider.ts [hunk 6]
```diff
@@ -245,6 +263,9 @@ export default class LokiLanguageProvider extends LanguageProvider {
         ...this.labelKeys,
         [EMPTY_SELECTOR]: labelKeys,
       };
+      this.labelValues = {
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/language_provider.ts]: This is the actual fix. Not sure if the best place but it seems like calling `start()` is required for things to function anyway.

**Review by kaydelaney — APPROVED** [review]
LGTM

**Pull request merged**